[quant][core][gpu][improvement] Converted reinterpret_cast<T *>(some_int8_tensor.data_ptr()) calls to some_int8_tensor.data_ptr<int8_t> in quantized cudnn operator files

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75980

Summary:
Support for data_ptr<T> for quantized tensor was enabled in
https://github.com/pytorch/pytorch/pull/75643. Rather than using
reinterpret_cast, we can use this overload directly. The change is
currently made in /aten/src/ATen/native/quantized/cudnn.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
python test/test_quantization.py -k test_qconv2d_cudnn
python test/test_quantization.py -k test_qadd_relu_cudnn
```

Differential Revision: [D35720654](https://our.internmc.facebook.com/intern/diff/D35720654)

dzdang committed:
[quant][core][gpu][improvement] Converted reinterpret_cast<T *>(some_int8_tensor.data_ptr()) calls to some_int8_tensor.data_ptr<int8_t> in quantized cudnn operator files

Summary:
Support for data_ptr<T> for quantized tensor was enabled in
https://github.com/pytorch/pytorch/pull/75643. Rather than using
reinterpret_cast, we can use this overload directly. The change is
currently made in /aten/src/ATen/native/quantized/cudnn.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
python test/test_quantization.py -k test_qconv2d_cudnn
python test/test_quantization.py -k test_qadd_relu_cudnn
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35720654).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Conv.cpp:
```diff
@@ -146,10 +146,8 @@ void PackedConvWeightCudnn<kSpatialDim>::apply_impl_helper(const at::Tensor& qua
-                                           requantize_multiplier_tensor.data_ptr(),
-                                           reinterpret_cast<int8_t*>(quantized_output.data_ptr())};
+    data_ptrs = {input.data_ptr<int8_t>(), conv_output.data_ptr(), orig_weight_.data_ptr<int8_t>(),
+                 requantize_multiplier_tensor.data_ptr(), quantized_output.data_ptr<int8_t>()};
```
nit: should we make this more explicit as well

dzdang committed:
Update on "[quant][core][gpu][improvement] Converted reinterpret_cast<T *>(some_int8_tensor.data_ptr()) calls to some_int8_tensor.data_ptr<int8_t> in quantized cudnn operator files"


Summary:
Support for data_ptr<T> for quantized tensor was enabled in
https://github.com/pytorch/pytorch/pull/75643. Rather than using
reinterpret_cast, we can use this overload directly. The change is
currently made in /aten/src/ATen/native/quantized/cudnn.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
python test/test_quantization.py -k test_qconv2d_cudnn
python test/test_quantization.py -k test_qadd_relu_cudnn
```

Differential Revision: [D35720654](https://our.internmc.facebook.com/intern/diff/D35720654)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35720654).
